janusgraph-core/src/main/java/org/janusgraph/graphdb/olap/job/IndexRepairJob.java

```diff
@@ -124,20 +125,48 @@ public void process(JanusGraphVertex vertex, ScanMetrics metrics) {
                 RelationTypeIndexWrapper wrapper = (RelationTypeIndexWrapper)index;
                 InternalRelationType wrappedType = wrapper.getWrappedType();
                 EdgeSerializer edgeSerializer = writeTx.getEdgeSerializer();
-                List<Entry> additions = new ArrayList<>();
-
+                List<Entry> outAdditions = new ArrayList<>();
+                Map<StaticBuffer,List<Entry>> inAdditionsMap = new HashMap<>();
+                
                 for (Object relation : vertex.query().types(indexRelationTypeName).direction(Direction.OUT).relations()) {
                     InternalRelation janusgraphRelation = (InternalRelation) relation;
                     for (int pos = 0; pos < janusgraphRelation.getArity(); pos++) {
                         if (!wrappedType.isUnidirected(Direction.BOTH) && !wrappedType.isUnidirected(EdgeDirection.fromPosition(pos)))
                             continue; //Directionality is not covered
+
                         Entry entry = edgeSerializer.writeRelation(janusgraphRelation, wrappedType, pos, writeTx);
-                        additions.add(entry);
+                        
+                        //The below condition is check to avoid self-links which is getting created after re-indexing
+                        if(pos==0) {
+                            //Create OUT edge index entry. 
+                            //Here source will the current vertex and target vertex the other side of the relation 
+                            outAdditions.add(entry); 
+                        }else if(pos==1) {
+                            //Create IN edge index entry.
+                            //Here the source vertex is the other side of the current vertex and target will be the current vertex
+                            InternalVertex otherVertex = janusgraphRelation.getVertex(1);
+                            StaticBuffer otherVertexKey = writeTx.getIdInspector().getKey(otherVertex.longId());
+                            inAdditionsMap.computeIfAbsent(otherVertexKey, k -> new ArrayList<Entry>()).add(entry);
```

Comment:
I find that your test case testReindexingForEdgeIndex will pass even if I remove this whole "else" branch. There are two possibilities:
1) Your test case is not effective enough to cover the edge case
2) This is actually not needed. Maybe only the case where pos == 0 is needed.
Can you please investigate?

Reply:
@li-boxuan  Sure i ll look into this issue.

Reply:
Your fix is correct. It's just that your test case is not effective enough. I can provide a subsequent PR to improve the tests.